# Sentiment Analysis on Stocks from Daily News

### Objective
Build a model that analyse the sentiment from the top 25 headlines for a stock and then predict whether it will go up or down.

### Data
https://www.kaggle.com/aaron7sun/stocknews

In [1]:
# standard imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# configurations
%matplotlib inline
sns.set_style('whitegrid')
pd.set_option("display.max_rows", 1000)
pd.set_option("display.max_columns", 1000)

# supress warnings
import warnings
warnings.filterwarnings('ignore')

In [3]:
# custom feature names
names = ['Date', 'Label', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15',
       '16', '17', '18', '19', '20', '21', '22', '23', '24', '25']

In [4]:
# load dataset
df = pd.read_csv("Combined_News_DJIA.csv", encoding='ISO-8859-1', names=names)

# Explore dataset

In [5]:
# preview dataset
df.head()

Date  Label                                                  1  \
0        Date  Label                                               Top1   
1  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
2  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
3  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   
4  2008-08-13      0  b' U.S. refuses Israel weapons to attack Iran:...   

                                                   2  \
0                                               Top2   
1            b'BREAKING: Musharraf to be impeached.'   
2        b'Bush puts foot down on Georgian conflict'   
3                 b"Russia 'ends Georgia operation'"   
4  b"When the president ordered to attack Tskhinv...   

                                                   3  \
0                                               Top3   
1  b'Russia Today: Columns of troops roll into So...   
2  b"Jewish Georgian minister: Thanks to Israeli ...   
3  b'"If we had no sexual harassment we would hav...   
4  b' Israel clears troops who killed Reuters cam...   

                                                   4  \
0                                               Top4   
1  b'Russian tanks are moving towards the capital...   
2  b'Georgian army flees in disarray as Russians ...   
3  b"Al-Qa'eda is losing support in Iraq because ...   
4  b'Britain\'s policy of being tough on drugs is...   

                                                   5  \
0                                               Top5   
1  b"Afghan children raped with 'impunity,' U.N. ...   
2      b"Olympic opening ceremony fireworks 'faked'"   
3  b'Ceasefire in Georgia: Putin Outmaneuvers the...   
4  b'Body of 14 year old found in trunk; Latest (...   

                                                   6  \
0                                               Top6   
1  b'150 Russian tanks have entered South Ossetia...   
2  b'What were the Mossad with fraudulent New Zea...   
3  b'Why Microsoft and Intel tried to kill the XO...   
4  b'China has moved 10 *million* quake survivors...   

                                                   7  \
0                                               Top7   
1  b"Breaking: Georgia invades South Ossetia, Rus...   
2  b'Russia angered by Israeli military sale to G...   
3  b'Stratfor: The Russo-Georgian War and the Bal...   
4  b"Bush announces Operation Get All Up In Russi...   

                                                   8  \
0                                               Top8   
1  b"The 'enemy combatent' trials are nothing but...   
2  b'An American citizen living in S.Ossetia blam...   
3  b"I'm Trying to Get a Sense of This Whole Geor...   
4             b'Russian forces sink Georgian ships '   

                                                   9  \
0                                               Top9   
1  b'Georgian troops retreat from S. Osettain cap...   
2  b'Welcome To World War IV! Now In High Definit...   
3  b"The US military was surprised by the timing ...   
4  b"The commander of a Navy air reconnaissance s...   

                                                  10  \
0                                              Top10   
1  b'Did the U.S. Prep Georgia for War with Russia?'   
2  b"Georgia's move, a mistake of monumental prop...   
3    b'U.S. Beats War Drum as Iran Dumps the Dollar'   
4  b"92% of CNN readers: Russia's actions in Geor...   

                                                  11  \
0                                              Top11   
1  b'Rice Gives Green Light for Israel to Attack ...   
2  b'Russia presses deeper into Georgia; U.S. say...   
3  b'Gorbachev: "Georgian military attacked the S...   
4  b'USA to send fleet into Black Sea to help Geo...   

                                                  12  \
0                                              Top12   
1  b'Announcing:Class Action Lawsuit on Behalf of...   
2  b'Abhinav Bindra 

In [6]:
df.tail()

Date Label                                                  1  \
1985  2016-06-27     0  Barclays and RBS shares suspended from trading...   
1986  2016-06-28     1  2,500 Scientists To Australia: If You Want To ...   
1987  2016-06-29     1                   Explosion At Airport In Istanbul   
1988  2016-06-30     1  Jamaica proposes marijuana dispensers for tour...   
1989  2016-07-01     1  A 117-year-old woman in Mexico City finally re...   

                                                      2  \
1985  Pope says Church should ask forgiveness from g...   
1986  The personal details of 112,000 French police ...   
1987  Yemeni former president: Terrorism is the offs...   
1988  Stephen Hawking says pollution and 'stupidity'...   
1989   IMF chief backs Athens as permanent Olympic host   

                                                      3  \
1985  Poland 'shocked' by xenophobic abuse of Poles ...   
1986  S&amp;P cuts United Kingdom sovereign credit r...   
1987  UK must accept freedom of movement to access E...   
1988  Boris Johnson says he will not run for Tory pa...   
1989  The president of France says if Brexit won, so...   

                                                      4  \
1985  There will be no second referendum, cabinet ag...   
1986                Huge helium deposit found in Africa   
1987  Devastated: scientists too late to captive bre...   
1988  Six gay men in Ivory Coast were abused and for...   
1989  British Man Who Must Give Police 24 Hours' Not...   

                                                      5  \
1985      Scotland welcome to join EU, Merkel ally says   
1986  CEO of the South African state broadcaster qui...   
1987  British Labor Party leader Jeremy Corbyn loses...   
1988  Switzerland denies citizenship to Muslim immig...   
1989  100+ Nobel laureates urge Greenpeace to stop o...   

                                                      6  \
1985  Sterling dips below Friday's 31-year low amid ...   
1986  Brexit cost investors $2 trillion, the worst o...   
1987  A Muslim Shop in the UK Was Just Firebombed Wh...   
1988  Palestinian terrorist stabs israeli teen girl ...   
1989  Brazil: Huge spike in number of police killing...   

                                                      7  \
1985  No negative news about South African President...   
1986  Hong Kong democracy activists call for return ...   
1987  Mexican Authorities Sexually Torture Women in ...   
1988  Puerto Rico will default on $1 billion of debt...   
1989  Austria's highest court annuls presidential el...   

                                                      8  \
1985  Surge in Hate Crimes in the U.K. Following U.K...   
1986  Brexit: Iceland president says UK can join 'tr...   
1987            UK shares and pound continue to recover   
1988  Republic of Ireland fans to be awarded medal f...   
1989  Facebook wins privacy case, can track any Belg...   

                                                      9  \
1985  Weapons shipped into Jordan by the CIA and Sau...   
1986  UK's Osborne: 'Absolutely' going to have to cu...   
1987  Iceland historian Johannesson wins presidentia...   
1988  Afghan suicide bomber 'kills up to 40' - BBC News   
1989  Switzerland denies Muslim girls citizenship af...   

                                                     10  \
1985  Angela Merkel said the U.K. must file exit pap...   
1986  'Do not let Scotland down now' : Scottish MEP ...   
1987  99-Million-Yr-Old Bird Wings Found Encased in ...   
1988  US airstrikes kill at least 250 ISIS fighters ...   
1989  China kills millions of innocent meditators fo...   

                                                     11  \
1985  In a birth offering hope to a threatened speci...   
1986  British pound could hit history-making dollar ...   
1987  A chatbot programmed by a British teenager has...   
1988  Turkish Cop Who Took Down Istanbul Gunman Hail...   
1989  France Cracks Down on Factory Farms - A viral ...   


In [7]:
# dataset size
print("# rows:",df.shape[0])
print("# cols:", df.shape[1])

# rows: 1990
# cols: 27


In [8]:
# check for null values
df.isnull().sum()

Date     0
Label    0
1        0
2        0
3        0
4        0
5        0
6        0
7        0
8        0
9        0
10       0
11       0
12       0
13       0
14       0
15       0
16       0
17       0
18       0
19       0
20       0
21       0
22       0
23       1
24       3
25       3
dtype: int64

- very few data records missing in the dataset that we will treat in future

# Train Test Split

In [9]:
# split dataset into train and test datasets
train = df[df['Date'] < '20150101']
test = df[df['Date'] > '20141231']

In [10]:
# size of datasets
print("Train Dataset")
print("# rows:", train.shape[0])
print("# cols:", train.shape[1])
print()
# size of datasets
print("Test Dataset")
print("# rows:", test.shape[0])
print("# cols:", test.shape[1])

Train Dataset
# rows: 1863
# cols: 27

Test Dataset
# rows: 379
# cols: 27


# Text Preprocessing

In [11]:
# drop date and target variable from train dataset
data=train.iloc[:,2:27]

# remove punctuations
data.replace("[^a-zA-Z]"," ",regex=True, inplace=True)

In [12]:
# convert all words into lowercase
for i in data.columns:
    data[i] = data[i].str.lower()

In [13]:
# preview dataset
data.head(3)

1  \
1  b georgia  downs two russian warplanes  as cou...   
2  b why wont america and nato help us  if they w...   
3  b remember that adorable   year old who sang a...   

                                             2  \
1      b breaking  musharraf to be impeached     
2  b bush puts foot down on georgian conflict    
3           b russia  ends georgia operation     

                                                   3  \
1  b russia today  columns of troops roll into so...   
2  b jewish georgian minister  thanks to israeli ...   
3  b  if we had no sexual harassment we would hav...   

                                                   4  \
1  b russian tanks are moving towards the capital...   
2  b georgian army flees in disarray as russians ...   
3  b al qa eda is losing support in iraq because ...   

                                                   5  \
1  b afghan children raped with  impunity   u n  ...   
2      b olympic opening ceremony fireworks  faked     
3  b ceasefire in georgia  putin outmaneuvers the...   

                                                   6  \
1  b     russian tanks have entered south ossetia...   
2  b what were the mossad with fraudulent new zea...   
3  b why microsoft and intel tried to kill the xo...   

                                                   7  \
1  b breaking  georgia invades south ossetia  rus...   
2  b russia angered by israeli military sale to g...   
3  b stratfor  the russo georgian war and the bal...   

                                                   8  \
1  b the  enemy combatent  trials are nothing but...   
2  b an american citizen living in s ossetia blam...   
3  b i m trying to get a sense of this whole geor...   

                                                   9  \
1  b georgian troops retreat from s  osettain cap...   
2  b welcome to world war iv  now in high definit...   
3  b the us military was surprised by the timing ...   

                                                  10  \
1  b did the u s  prep georgia for war with russia     
2  b georgia s move  a mistake of monumental prop...   
3    b u s  beats war drum as iran dumps the dollar    

                                                  11  \
1  b rice gives green light for israel to attack ...   
2  b russia presses deeper into georgia  u s  say...   
3  b gorbachev   georgian military attacked the s...   

                                                  12  \
1  b announcing class action lawsuit on behalf of...   
2  b abhinav bindra wins first ever individual ol...   
3  b cnn use footage of tskhinvali ruins to cover...   

                                                  13  \
1  b so   russia and georgia are at war and the n...   
2  b  u s  ship heads for arctic to define territ...   
3  b beginning a war as the olympics were opening...   

                                                  14  \
1  b china tells bush to stay out of other countr...   
2  b drivers in a jerusalem taxi station threaten...   
3  b    pyramids as large as the luxor stacked in...   

                                                  15  \
1                  b did world war iii start today     
2  b the french team is stunned by phelps and the...   
3            b the    top party cities in the world    

                                                  16  \
1  b georgia invades south ossetia   if russia ge...   
2  b israel and the us behind the georgian aggres...   
3  b u s  troops still in georgia  did you know t...   

                                                  17  \
1                b al qaeda faces islamist backlash    
2  b  do not believe tv  neither russian nor geor...   
3       b why russias response to georgia was right    

                                                  18  \
1  b condoleezza rice   the us would not act to p...   
2  b riots are still going on in montreal  canada...   
3  b gorbachev accuses u s  of making a  serious ...   

                         

In [14]:
# combine all headlines

'''
# combine first row
' '.join(str(x) for x in data.iloc[0,0:25])
'''

headlines = []
for row in range(0, len(data.index)):
    headlines.append(" ".join(str(x) for x in data.iloc[row, 0:25]))

In [15]:
headlines

['b georgia  downs two russian warplanes  as countries move to brink of war  b breaking  musharraf to be impeached   b russia today  columns of troops roll into south ossetia  footage from fighting  youtube   b russian tanks are moving towards the capital of south ossetia  which has reportedly been completely destroyed by georgian artillery fire  b afghan children raped with  impunity   u n  official says   this is sick  a three year old was raped and they do nothing  b     russian tanks have entered south ossetia whilst georgia shoots down two russian jets   b breaking  georgia invades south ossetia  russia warned it would intervene on so s side  b the  enemy combatent  trials are nothing but a sham  salim haman has been sentenced to       years  but will be kept longer anyway just because they feel like it   b georgian troops retreat from s  osettain capital  presumably leaving several hundred people killed   video   b did the u s  prep georgia for war with russia   b rice gives gree

In [16]:
len(headlines)

1863

In [17]:
# all headlines are stored in a lists
type(headlines)

list

# Text Vectorization & Modelling

## Count Vectorizer & Random Forest

In [40]:
# import vectorizer and classifier
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier

In [44]:
# bag of words
vectorizer = CountVectorizer(ngram_range=(2,2))
train_data = vectorizer.fit_transform(headlines)

In [45]:
# implement RandomForestClassifier
randomForestClassifier = RandomForestClassifier(n_estimators=200, criterion='entropy')
randomForestClassifier.fit(train_data, train['Label'])

RandomForestClassifier(criterion='entropy', n_estimators=200)

In [46]:
# make predictions on test dataset
test_transform = []
for row in range(0, len(test.index)):
    test_transform.append(" ".join(str(x) for x in test.iloc[row, 2:27]))
    
test_dataset = vectorizer.transform(test_transform)
predictions = randomForestClassifier.predict(test_dataset)

In [47]:
predictions

array(['1', '1', '0', '0', '1', '1', '0', '0', '0', '0', '0', '1', '1',
       '1', '1', '0', '1', '0', '0', '1', '0', '1', '1', '1', '1', '0',
       '0', '1', '0', '1', '1', '1', '0', '0', '1', '0', '1', '1', '0',
       '0', '1', '0', '0', '1', '0', '1', '0', '0', '1', '0', '1', '0',
       '1', '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1',
       '0', '1', '1', '1', '0', '1', '1', '0', '0', '1', '0', '1', '1',
       '1', '0', '1', '0', '0', '1', '1', '0', '0', '1', '1', '0', '0',
       '0', '1', '1', '1', '1', '0', '1', '0', '0', '1', '0', '0', '1',
       '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1', '1',
       '0', '1', '1', '0', '0', '1', '0', '1', '1', '0', '0', '1', '0',
       '1', '1', '1', '1', '0', '1', '0', '1', '0', '0', '0', '0', '0',
       '1', '1', '0', '0', '0', '0', '0', '0', '0', '1', '0', '0', '1',
       '1', '1', '0', '0', '0', '0', '0', '0', '1', '1', '0', '0', '0',
       '1', '1', '0', '1', '0', '1', '1', '0', '1', '1', '0', '0

In [48]:
print(test['Label'])

0       Label
1612        1
1613        0
1614        0
1615        1
1616        1
1617        0
1618        0
1619        0
1620        0
1621        0
1622        1
1623        1
1624        1
1625        1
1626        0
1627        1
1628        0
1629        0
1630        1
1631        0
1632        1
1633        1
1634        1
1635        1
1636        0
1637        0
1638        1
1639        0
1640        1
1641        1
1642        1
1643        0
1644        0
1645        1
1646        0
1647        1
1648        1
1649        0
1650        0
1651        1
1652        0
1653        0
1654        1
1655        0
1656        1
1657        0
1658        0
1659        1
1660        0
1661        1
1662        0
1663        1
1664        0
1665        1
1666        0
1667        0
1668        0
1669        0
1670        1
1671        1
1672        0
1673        0
1674        1
1675        1
1676        0
1677        1
1678        1
1679        1
1680        0
1681        1
1682  

## Model evaluation

In [49]:
# import packages
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [50]:
# accuracy
score = accuracy_score(test['Label'], predictions)
print("Accuracy", round(score,2)*100)
print()
# confusion matrix
matrix = confusion_matrix(test['Label'], predictions)
print(matrix)
print()

# report
report = classification_report(test['Label'], predictions)
print(report)

Accuracy 85.0

[[133  53   0]
 [  2 190   0]
 [  0   1   0]]

              precision    recall  f1-score   support

           0       0.99      0.72      0.83       186
           1       0.78      0.99      0.87       192
       Label       0.00      0.00      0.00         1

    accuracy                           0.85       379
   macro avg       0.59      0.57      0.57       379
weighted avg       0.88      0.85      0.85       379



## TFIDF Vectorizer + Random Forest

In [52]:
# import vectorizer and classifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier

In [53]:
# bag of words
vectorizer = TfidfVectorizer(ngram_range=(2,2))
train_data = vectorizer.fit_transform(headlines)

In [54]:
# implement RandomForestClassifier
randomForestClassifier = RandomForestClassifier(n_estimators=200, criterion='entropy')
randomForestClassifier.fit(train_data, train['Label'])

RandomForestClassifier(criterion='entropy', n_estimators=200)

In [55]:
# make predictions on test dataset
test_transform = []
for row in range(0, len(test.index)):
    test_transform.append(" ".join(str(x) for x in test.iloc[row, 2:27]))
    
test_dataset = vectorizer.transform(test_transform)
predictions = randomForestClassifier.predict(test_dataset)

In [56]:
predictions

array(['1', '1', '0', '0', '1', '1', '0', '0', '0', '0', '0', '1', '1',
       '1', '1', '0', '1', '0', '0', '1', '0', '1', '1', '1', '1', '0',
       '0', '1', '0', '1', '1', '1', '0', '0', '1', '0', '1', '1', '0',
       '0', '1', '0', '0', '1', '0', '1', '0', '0', '1', '0', '1', '0',
       '1', '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1',
       '0', '1', '1', '1', '0', '1', '1', '0', '0', '1', '0', '1', '1',
       '1', '0', '1', '0', '0', '1', '1', '0', '0', '1', '1', '0', '0',
       '0', '1', '1', '1', '1', '0', '1', '0', '0', '1', '0', '0', '1',
       '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1', '1',
       '0', '1', '1', '0', '0', '1', '0', '1', '1', '0', '0', '1', '0',
       '1', '1', '1', '1', '0', '1', '0', '1', '0', '0', '0', '0', '0',
       '1', '1', '0', '0', '0', '0', '0', '0', '0', '1', '0', '0', '1',
       '1', '1', '0', '0', '0', '0', '0', '0', '1', '1', '0', '0', '0',
       '1', '1', '0', '1', '0', '1', '1', '0', '1', '1', '0', '0

In [57]:
print(test['Label'])

0       Label
1612        1
1613        0
1614        0
1615        1
1616        1
1617        0
1618        0
1619        0
1620        0
1621        0
1622        1
1623        1
1624        1
1625        1
1626        0
1627        1
1628        0
1629        0
1630        1
1631        0
1632        1
1633        1
1634        1
1635        1
1636        0
1637        0
1638        1
1639        0
1640        1
1641        1
1642        1
1643        0
1644        0
1645        1
1646        0
1647        1
1648        1
1649        0
1650        0
1651        1
1652        0
1653        0
1654        1
1655        0
1656        1
1657        0
1658        0
1659        1
1660        0
1661        1
1662        0
1663        1
1664        0
1665        1
1666        0
1667        0
1668        0
1669        0
1670        1
1671        1
1672        0
1673        0
1674        1
1675        1
1676        0
1677        1
1678        1
1679        1
1680        0
1681        1
1682  

# Model evaluation

In [58]:
# import packages
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [59]:
# accuracy
score = accuracy_score(test['Label'], predictions)
print("Accuracy", round(score,2)*100)
print()
# confusion matrix
matrix = confusion_matrix(test['Label'], predictions)
print(matrix)
print()

# report
report = classification_report(test['Label'], predictions)
print(report)

Accuracy 85.0

[[148  38   0]
 [ 18 174   0]
 [  0   1   0]]

              precision    recall  f1-score   support

           0       0.89      0.80      0.84       186
           1       0.82      0.91      0.86       192
       Label       0.00      0.00      0.00         1

    accuracy                           0.85       379
   macro avg       0.57      0.57      0.57       379
weighted avg       0.85      0.85      0.85       379



- Accuracy is same in both vectorizers

## TFIDF Vectorizer + Naive Bayes

In [62]:
# import library
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()

# train
nb.fit(train_data, train['Label'])

MultinomialNB()

In [66]:
# make predictions on test dataset
test_transform = []
for row in range(0, len(test.index)):
    test_transform.append(" ".join(str(x) for x in test.iloc[row, 2:27]))
    
test_dataset = vectorizer.transform(test_transform)
predictions = nb.predict(test_dataset)

In [67]:
predictions

array(['1', '1', '0', '0', '1', '1', '0', '0', '0', '0', '0', '1', '1',
       '1', '1', '0', '1', '0', '0', '1', '0', '1', '1', '1', '1', '0',
       '0', '1', '0', '1', '1', '1', '0', '0', '1', '0', '1', '1', '0',
       '0', '1', '0', '0', '1', '0', '1', '0', '0', '1', '0', '1', '0',
       '1', '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1',
       '0', '1', '1', '1', '0', '1', '1', '0', '0', '1', '0', '1', '1',
       '1', '0', '1', '0', '0', '1', '1', '0', '0', '1', '1', '0', '0',
       '0', '1', '1', '1', '1', '0', '1', '0', '0', '1', '0', '0', '1',
       '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1', '1',
       '0', '1', '1', '0', '0', '1', '0', '1', '1', '0', '0', '1', '0',
       '1', '1', '1', '1', '0', '1', '0', '1', '0', '0', '0', '0', '0',
       '1', '1', '0', '0', '0', '0', '0', '0', '0', '1', '0', '0', '1',
       '1', '1', '0', '0', '0', '0', '0', '0', '1', '1', '0', '0', '0',
       '1', '1', '0', '1', '0', '1', '1', '0', '1', '1', '0', '0

In [68]:
test['Label']

0       Label
1612        1
1613        0
1614        0
1615        1
1616        1
1617        0
1618        0
1619        0
1620        0
1621        0
1622        1
1623        1
1624        1
1625        1
1626        0
1627        1
1628        0
1629        0
1630        1
1631        0
1632        1
1633        1
1634        1
1635        1
1636        0
1637        0
1638        1
1639        0
1640        1
1641        1
1642        1
1643        0
1644        0
1645        1
1646        0
1647        1
1648        1
1649        0
1650        0
1651        1
1652        0
1653        0
1654        1
1655        0
1656        1
1657        0
1658        0
1659        1
1660        0
1661        1
1662        0
1663        1
1664        0
1665        1
1666        0
1667        0
1668        0
1669        0
1670        1
1671        1
1672        0
1673        0
1674        1
1675        1
1676        0
1677        1
1678        1
1679        1
1680        0
1681        1
1682  

In [69]:
# accuracy
score = accuracy_score(test['Label'], predictions)
print("Accuracy", round(score,2)*100)
print()
# confusion matrix
matrix = confusion_matrix(test['Label'], predictions)
print(matrix)
print()

# report
report = classification_report(test['Label'], predictions)
print(report)

Accuracy 85.0

[[130  56   0]
 [  0 192   0]
 [  0   1   0]]

              precision    recall  f1-score   support

           0       1.00      0.70      0.82       186
           1       0.77      1.00      0.87       192
       Label       0.00      0.00      0.00         1

    accuracy                           0.85       379
   macro avg       0.59      0.57      0.56       379
weighted avg       0.88      0.85      0.84       379



## Count Vectorizer + Naive Bayes

In [70]:
# import vectorizer
from sklearn.feature_extraction.text import CountVectorizer

In [71]:
# bag of words
vectorizer = TfidfVectorizer(ngram_range=(2,2))
train_data = vectorizer.fit_transform(headlines)

In [ ]:
# import library
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()

# train
nb.fit(train_data, train['Label'])

In [72]:
# make predictions on test dataset
test_transform = []
for row in range(0, len(test.index)):
    test_transform.append(" ".join(str(x) for x in test.iloc[row, 2:27]))
    
test_dataset = vectorizer.transform(test_transform)
predictions = nb.predict(test_dataset)

In [73]:
predictions

array(['1', '1', '0', '0', '1', '1', '0', '0', '0', '0', '0', '1', '1',
       '1', '1', '0', '1', '0', '0', '1', '0', '1', '1', '1', '1', '0',
       '0', '1', '0', '1', '1', '1', '0', '0', '1', '0', '1', '1', '0',
       '0', '1', '0', '0', '1', '0', '1', '0', '0', '1', '0', '1', '0',
       '1', '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1',
       '0', '1', '1', '1', '0', '1', '1', '0', '0', '1', '0', '1', '1',
       '1', '0', '1', '0', '0', '1', '1', '0', '0', '1', '1', '0', '0',
       '0', '1', '1', '1', '1', '0', '1', '0', '0', '1', '0', '0', '1',
       '0', '1', '0', '0', '0', '0', '1', '1', '0', '0', '1', '1', '1',
       '0', '1', '1', '0', '0', '1', '0', '1', '1', '0', '0', '1', '0',
       '1', '1', '1', '1', '0', '1', '0', '1', '0', '0', '0', '0', '0',
       '1', '1', '0', '0', '0', '0', '0', '0', '0', '1', '0', '0', '1',
       '1', '1', '0', '0', '0', '0', '0', '0', '1', '1', '0', '0', '0',
       '1', '1', '0', '1', '0', '1', '1', '0', '1', '1', '0', '0

In [74]:
test['Label']

0       Label
1612        1
1613        0
1614        0
1615        1
1616        1
1617        0
1618        0
1619        0
1620        0
1621        0
1622        1
1623        1
1624        1
1625        1
1626        0
1627        1
1628        0
1629        0
1630        1
1631        0
1632        1
1633        1
1634        1
1635        1
1636        0
1637        0
1638        1
1639        0
1640        1
1641        1
1642        1
1643        0
1644        0
1645        1
1646        0
1647        1
1648        1
1649        0
1650        0
1651        1
1652        0
1653        0
1654        1
1655        0
1656        1
1657        0
1658        0
1659        1
1660        0
1661        1
1662        0
1663        1
1664        0
1665        1
1666        0
1667        0
1668        0
1669        0
1670        1
1671        1
1672        0
1673        0
1674        1
1675        1
1676        0
1677        1
1678        1
1679        1
1680        0
1681        1
1682  

In [75]:
# accuracy
score = accuracy_score(test['Label'], predictions)
print("Accuracy", round(score,2)*100)
print()
# confusion matrix
matrix = confusion_matrix(test['Label'], predictions)
print(matrix)
print()

# report
report = classification_report(test['Label'], predictions)
print(report)

Accuracy 85.0

[[130  56   0]
 [  0 192   0]
 [  0   1   0]]

              precision    recall  f1-score   support

           0       1.00      0.70      0.82       186
           1       0.77      1.00      0.87       192
       Label       0.00      0.00      0.00         1

    accuracy                           0.85       379
   macro avg       0.59      0.57      0.56       379
weighted avg       0.88      0.85      0.84       379



# Conclusion

- All models are giving us an accuracy of around 85%
- Therefore, we can now select the one that gives us better predictions that can be evaluated from the confusion matrix